lib/ssoe/service.rb

```diff
@@ -56,5 +55,61 @@ def build_message(credential_method, credential_id, user, address)
         zipcode: address.zipcode
       ).perform
     end
+
+    def parse_response(response_body)
+      parsed = Hash.from_xml(response_body)
+
+      if parsed.dig('Envelope', 'Body', 'getSSOeTraitsByCSPIDResponse', 'icn')
+        icn = parsed['Envelope']['Body']['getSSOeTraitsByCSPIDResponse']['icn']
+        return { success: true, icn: } if icn.present?
+      end
+
+      if parsed.dig('Envelope', 'Body', 'Fault')
+        fault_code = parsed.dig('Envelope', 'Body', 'Fault', 'faultcode')
+        fault_string = parsed.dig('Envelope', 'Body', 'Fault', 'faultstring')
+
+        return {
+          success: false,
+          error: {
+            code: fault_code || 'UnknownError',
+            message: fault_string || 'Unable to parse SOAP response'
+          }
+        }
+      end
+
+      unknown_error
```

Comment:
Instead of defining a method to handle unknown errors and using it twice, I would update the end of this parsing logic to throw some sort of error that would then get caught by the error handling that is calling the same `unknown_error`; you could then dispense with the method and just have the error handling in that one `rescue`.

I'd also consider moving this `rescue` out of the `parse_response` method and into the the parent `get_traits` method that is actually being called - it's best to have your error handling in the top level methods in a class because those are the ones that are being interacted with by the rest of your application.